python/pylibraft/pylibraft/distance/pairwise_distance.pyx

```diff
@@ -0,0 +1,135 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+import numpy as np
+
+from libc.stdint cimport uintptr_t
+from cython.operator cimport dereference as deref
+
+from libcpp cimport bool
+from .distance_type cimport DistanceType
+from pylibraft.common.handle cimport handle_t
+
+cdef extern from "raft_distance/pairwise_distance.hpp" \
+        namespace "raft::distance::runtime":
+
+    cdef void pairwise_distance(const handle_t &handle,
+                                float *x,
+                                float *y,
+                                float *dists,
+                                int m,
+                                int n,
+                                int k,
+                                DistanceType metric,
+                                bool isRowMajor,
+                                float metric_arg)
+
+    cdef void pairwise_distance(const handle_t &handle,
+                                double *x,
+                                double *y,
+                                double *dists,
+                                int m,
+                                int n,
+                                int k,
+                                DistanceType metric,
+                                bool isRowMajor,
+                                float metric_arg)
+
+DISTANCE_TYPES = {
+    "l2": DistanceType.L2SqrtUnexpanded,
+    "euclidean": DistanceType.L2SqrtUnexpanded,
+    "l1": DistanceType.L1,
+    "cityblock": DistanceType.L1,
+    "inner_product": DistanceType.InnerProduct,
+    "chebyshev": DistanceType.Linf,
+    "canberra": DistanceType.Canberra,
+    "lp": DistanceType.LpUnexpanded,
+    "correlation": DistanceType.CorrelationExpanded,
+    "jaccard": DistanceType.JaccardExpanded,
+    "hellinger": DistanceType.HellingerExpanded,
+    "braycurtis": DistanceType.BrayCurtis,
+    "jensenshannon": DistanceType.JensenShannon,
+    "hamming": DistanceType.HammingUnexpanded,
+    "kl_divergence": DistanceType.KLDivergence,
+    "russellrao": DistanceType.RusselRaoExpanded,
+    "dice": DistanceType.DiceExpanded
+}
+
+SUPPORTED_DISTANCES = list(DISTANCE_TYPES.keys())
+
+
+def distance(X, Y, dists, metric="euclidean"):
+    """
+    Compute pairwise distances between X and Y
+
+    Parameters
+    ----------
+
+    X : CUDA array interface matrix shape (m, k)
+    Y : CUDA array interface matrix shape (n, k)
+    dists : Writable CUDA array interface matrix shape (m, n)
+    metric : string denoting the metric type
+    """
+
+    # TODO: Validate inputs, shapes, etc...
```

Comment:
Discussion for beyond this PR: since these are quite low level APIs (at least as they are right now), it might be worth to make the validation an option that can be disabled